src/solver.hs

```diff
@@ -0,0 +1,232 @@
+module Solver where
+
+import Sudoku
+import SudokuLogic
+import Sudokus
+import Optimizer
+import Data.List
+data Step =     LastFreeCellBlock Position      |
+                LastFreeCellRow Position        |
+                LastFreeCellCollumn Position    |
+                SingleCandidate Position        |
+                SinglePositionRow Position      |
+                SinglePositionColumn Position   |
+                SinglePositionBlock Position    |
+                CandidateLine Position          |
+                NakedPairRow Position           |
+                NakedPairColumn Position        |
+                NakedPairBlock Position         |
+                HiddenPairRow Position          |
+                HiddenPairColumn Position       |
+                HiddenPairBlock Position        |
+                NOAVAILABLESTEPS
```

Comment:
Maybe having a unified type of steps is a bit ambitious. Each technique may work a bit differently, so it may give you more freedom to have different functions for different techniques.